src/compiler/preprocess/index.ts

```diff
@@ -1,218 +1,168 @@
-		}
+		decoded_map = decode_map(processed);
+		
 		// offset only segments pointing at original component source
 		const source_index = decoded_map.sources.indexOf(file_basename);
 		if (source_index !== -1) {
-			sourcemap_add_offset(decoded_map, get_location(offset + prefix.length), source_index);
+			sourcemap_add_offset(decoded_map, location, source_index);
 		}
 	}
-	const processed_with_map = StringWithSourcemap.from_processed(processed.code, decoded_map);
 
-	// Surround the processed code with the prefix and suffix, retaining valid sourcemappings
-	return prefix_with_map.concat(processed_with_map).concat(suffix_with_map);
+	return StringWithSourcemap.from_processed(processed.code, decoded_map);
+}
+
+/**
+ * Given the whole tag including content, return a `StringWithSourcemap`
+ * representing the tag content replaced with `processed`.
+ */
+function processed_tag_to_sws(
+	processed: Processed,
+	tag_name: 'style' | 'script',
+	attributes: string,
+	{ source, file_basename, get_location }: Source): StringWithSourcemap {
+	const build_sws = (source: string, offset: number) =>
+		StringWithSourcemap.from_source(file_basename, source, get_location(offset));
+
+	const tag_open = `<${tag_name}${attributes || ''}>`;
+	const tag_close = `</${tag_name}>`;
+
+	const tag_open_sws = build_sws(tag_open, 0);
+	const tag_close_sws = build_sws(tag_close, tag_open.length + source.length);
+
+	const content_sws = processed_content_to_sws(processed, get_location(tag_open.length), file_basename);
+
+	return tag_open_sws.concat(content_sws).concat(tag_close_sws);
+}
+
+/** 
+ * Calculate the updates required to process all instances of the specified tag. 
+ */
+async function process_tag(
+	tag_name: 'style' | 'script',
+	preprocessor: Preprocessor,
+	source: Source
+): Promise<SourceUpdate> {
+	const { file_basename, filename, get_location } = source;
+	const tag_regex =
+		tag_name === 'style'
+			? /<!--[^]*?-->|<style(\s[^]*?)?(?:>([^]*?)<\/style>|\/>)/gi
+			: /<!--[^]*?-->|<script(\s[^]*?)?(?:>([^]*?)<\/script>|\/>)/gi;
+
+	const dependencies: string[] = [];
+
+	async function process_single_tag(
+		tag_with_content: string,
+		attributes = '',
+		content = '',
+		tag_offset: number
+	): Promise<StringWithSourcemap> {
+		const no_change = () =>
+			StringWithSourcemap.from_source(file_basename, tag_with_content, get_location(tag_offset));
+
+		if (!attributes && !content) return no_change();
+
+		const processed = await preprocessor({
+			content: content || '',
+			attributes: parse_tag_attributes(attributes || ''),
+			filename
+		});
+
+		if (processed && processed.dependencies) dependencies.push(...processed.dependencies);
+		if (!processed || !processed.map && processed.code === content) return no_change();
+
+		return processed_tag_to_sws(processed, tag_name, attributes,
+			{source: content, get_location: offset => source.get_location(offset + tag_offset), filename, file_basename});
```

Comment:
this feel like another `PreprocessResult`, except the location get offset.

```js
{source: content, get_location: offset => source.get_location(offset + tag_offset), filename, file_basename})
```

would it be better if `PreprocessResult` have a method that can clone itself with an offset ?

Reply:
That makes sense. I gave it a stab; I am not entirely happy with the outcome, but it does feel like a step in the right direction so I'll leave it in. 

I suspect more can be done but I'm a bit afraid of doing any change that is not guaranteed to do 100% the same thing as the existing code as I still can't quite overview what it does...